core/src/shred.rs

```diff
@@ -304,96 +300,126 @@ impl Shredder {
+                    .enumerate()
+                    .map(|(i, shred_data)| {
+                        let shred_index = next_shred_index + i as u32;
+
+                        let mut header: u8 = 0;
+                        if i == last_shred_index as usize {
+                            header |= DATA_COMPLETE_SHRED;
+                            if is_last_in_slot {
+                                header |= LAST_SHRED_IN_SLOT;
+                            }
+                        }
 
-    fn sign_unsigned_shreds_and_generate_codes(&mut self) {
-        let signature_offset = *SIZE_OF_CODING_SHRED_HEADER;
-        let signer = self.signer.clone();
-        PAR_THREAD_POOL.with(|thread_pool| {
+                        let mut shred = Shred::new_from_data(
+                            self.slot,
+                            shred_index,
+                            (self.slot - self.parent_slot) as u16,
+                            Some(shred_data),
+                            header,
+                        );
+
+                        Shredder::sign_shred(
+                            &self.keypair,
+                            &mut shred,
+                            *SIZE_OF_CODING_SHRED_HEADER,
+                        );
+                        shred
+                    })
+                    .collect()
+            })
+        });
+
+        // 2) Generate coding shreds
+        let mut coding_shreds: Vec<_> = PAR_THREAD_POOL.with(|thread_pool| {
             thread_pool.borrow().install(|| {
-                self.shreds[self.fec_set_shred_start..]
-                    .par_iter_mut()
-                    .for_each(|d| Self::sign_shred(&signer, d, signature_offset));
+                finalized_shreds
+                    .par_chunks(MAX_DATA_SHREDS_PER_FEC_BLOCK as usize)
+                    .flat_map(|shred_data_batch| {
+                        Shredder::generate_coding_shreds(self.slot, self.fec_rate, shred_data_batch)
+                    })
+                    .collect()
             })
         });
-        let unsigned_coding_shred_start = self.shreds.len();
-
-        if self.fec_rate > 0.0 {
-            self.generate_coding_shreds();
-            let signature_offset = *SIZE_OF_SHRED_TYPE;
-            PAR_THREAD_POOL.with(|thread_pool| {
-                thread_pool.borrow().install(|| {
-                    self.shreds[unsigned_coding_shred_start..]
-                        .par_iter_mut()
-                        .for_each(|d| Self::sign_shred(&signer, d, signature_offset));
-                })
-            });
-        } else {
-            self.fec_set_index = self.index;
-        }
-        self.fec_set_shred_start = self.shreds.len();
-    }
 
-    /// Finalize a data shred. Update the shred index for the next shred
-    fn finalize_data_shred(&mut self) {
-        self.active_offset = 0;
-        self.index += 1;
+        // 3) Sign coding shreds
+        PAR_THREAD_POOL.with(|thread_pool| {
```

Comment:
Instead of calling install on the threadpool 3 different times is there a way to structure this to call it once? You can par_iter within an install as much as you want. 
The rayon bug basically kicks in _after_ the installed closure is complete.

Reply:
hmm so I tried coalescing all 3 into the same `install` call, but that for some reason performs marginally worse :P

Reply:
Alright as long as you're confident in your data lol.